Code/GraphMol/Depictor/RDDepictor.cpp

```diff
@@ -451,57 +450,160 @@ unsigned int compute2DCoordsMimicDistMat(
+    if (referencePattern->getNumAtoms(true) > reference.getNumAtoms(true)) {
       throw RDDepict::DepictException(
-          "When a pattern is provided, it must have the same number of atoms "
-          "as the reference");
+          "When a pattern is provided, its number of atoms must be "
+          "<= reference");
     }
     RDKit::MatchVectType refMatchVect;
     RDKit::SubstructMatch(reference, *referencePattern, refMatchVect);
     if (refMatchVect.empty()) {
       throw RDDepict::DepictException(
           "Reference pattern does not map to reference.");
     }
-    refMatch.reserve(refMatchVect.size());
+    refMatch.resize(reference.getNumAtoms(true), -1);
     for (auto &i : refMatchVect) {
-      refMatch.push_back(i.second);
+      refMatch[i.first] = i.second;
     }
-    RDKit::SubstructMatch(mol, *referencePattern, matchVect);
   } else {
-    refMatch.reserve(reference.getNumAtoms(true));
-    for (unsigned int i = 0; i < reference.getNumAtoms(true); ++i) {
-      refMatch.push_back(i);
-    }
-    RDKit::SubstructMatch(mol, reference, matchVect);
+    refMatch.resize(reference.getNumAtoms(true));
+    std::iota(refMatch.begin(), refMatch.end(), 0);
   }
-
-  RDGeom::INT_POINT2D_MAP coordMap;
-  if (matchVect.empty()) {
-    if (!acceptFailure) {
-      throw RDDepict::DepictException(
-          "Substructure match with reference not found.");
-    }
-  } else {
-    const RDKit::Conformer &conf = reference.getConformer(confId);
-    for (RDKit::MatchVectType::const_iterator mv = matchVect.begin();
-         mv != matchVect.end(); ++mv) {
-      RDGeom::Point3D pt3 = conf.getAtomPos(refMatch[mv->first]);
-      RDGeom::Point2D pt2(pt3.x, pt3.y);
-      coordMap[mv->second] = pt2;
+  const RDKit::ROMol &query =
+      (referencePattern ? *referencePattern : reference);
+  if (allowRGroups) {
+    std::unique_ptr<RDKit::ROMol> molHs(RDKit::MolOps::addHs(mol));
+    if (molHs) {
+      detail::MatchScorer matchScorer(*molHs, query);
+      auto matches = SubstructMatch(*molHs, query);
+      if (matches.empty()) {
+        allowRGroups = false;
+      } else {
+        matchVect = *std::min_element(
+            matches.begin(), matches.end(),
+            [&matchScorer](const RDKit::MatchVectType &aMatch,
+                           const RDKit::MatchVectType &bMatch) {
+              return (matchScorer.score(aMatch) < matchScorer.score(bMatch));
+            });
+        if (matchScorer.score(matchVect) > 1.0) {
+          for (const auto &pair : matchVect) {
+            if (molHs->getAtomWithIdx(pair.second)->getAtomicNum() != 1) {
+              matchVectFiltered.push_back(pair);
+            }
+          }
+          matchVectRef = matchVectFiltered;
+        }
+      }
     }
   }
-  RDDepict::compute2DCoords(mol, &coordMap, false /* canonOrient */,
-                            true /* clearConfs */, 0, 0, 0, false, forceRDKit);
+  if (!allowRGroups) {
```

Comment:
No, because `allowRGroups` can be set to `false` in the previous `if` clause. That may happen in the unlikely (but not impossible case) that a nasty user has specified a SMARTS query with implicit Hs (lowercase `h`) that would fail against the molecules with explicit Hs, but would succeed against the original molecule with implicit hs.